consul-lambda-registrator/extension/extension.go

```diff
@@ -128,24 +148,46 @@ func (ext *Extension) runEvents(ctx context.Context, errChan chan error) {
 	errChan <- nil
 }
 
-func (ext *Extension) getExtensionData(ctx context.Context) (structs.ExtensionData, error) {
+// getExtensionData asynchronously retrieves the extension data and updates the local cache.
+func (ext *Extension) getExtensionData(ctx context.Context) error {
 	trace.Enter()
 	defer trace.Exit()
 
-	var extData structs.ExtensionData
+	// Lock the extension data mutex so that the proxy's dial func will wait for this
+	// update to complete before attempting to dial out.
+	ext.dataMutex.Lock()
```

Comment:
Do we need to lock this early? It would be nice to only lock if we are updating the data.

Reply:
I had another look at this and now I remember the reason I put the lock before `getExtensionData` is called. I used it as a synchronization mechanism: Prior to the initial fetch, `ext.data` will be empty and if we don't put the lock around `getExtensionData` then the `DialFunc` will immediately use the empty `ext.data`.

I think there's a way to handle this though so that we only block on `getExtensionData` the first time.

I considered using a channel for notifying consumers of `ext.data` updates but we still have the problem of the first fetch.. and using a channel seemed to make it more complicated.